Dev Notes: Add section on E2E tests (#610)
CATcher-org/CATcher: docs/developer-notes.md

One main guideline is that a `describe` block should be created for each method / scenario under test, and an `it` block should be created for each property being verified.

### E2E Testing
#### CATcher Test Prep
E2E Tests are currently run using [Protractor](http://www.protractortest.org/#/) testing framework with the following stages. 
1. Build CATcher with using `test` architecture.
   - Using `test` build configuration located in `angular.json` under `projects.catcher.architect.configurations` we build a version of CATcher within a test environment that replaces `src/environments/environment.ts` with `src/environments/environment.test.ts` on runtime. This allows for the feeding of information into the application to differ its actions in comparison to the default / production environments.
2. Provide Test Environment Information
   - The Test Environment (in `src/environments/environment.test.ts`) provides information such as,
     - Login Credentials (Username).
     - User Team and Tutorial Information.
     - "Test Mode" activation by setting `test` property to `true`.
3. Mock Service Injections

>    - Data reflected in the environment file then assists the application replace the use of some existing services with those that bypass specific functions that are irrelevant to E2E Testing. This includes Authentication Bypassing, Backend API Simulation (so that tests can be carried out in isolationj) and others.

Reviewer: ```suggestion
   - Data reflected in the environment file then assists the application to replace the use of some existing services with those that bypass specific functions that are irrelevant to E2E Testing. This includes Authentication Bypassing, Backend API Simulation (so that tests can be carried out in isolation) and others.
```

4. Browser Action Injections using Protractor
   - With the application ready for testing, we then utilize `Protractor` to devise test cases that are located in the `/e2e` directory.
#### Protractor Configuration
- Protractor primarily requires the `*.conf.js` files to define E2E Testing Environments (this includes Browser Details, Base URL, etc...)
- The base configuration data is stored in `protractor.base.conf.js` which is then extended by separate configuration files for individual browsers as well as the CI/CD pipeline.
- E2E Tests are typically split into `Page-Objects Files` and `Test Files` in accordance with the [Protractor Style Guide](http://www.protractortest.org/#/style-guide) (more information regarding the interaction between the aforementioned filetypes can be found there).
- E2E Tests are also grouped into suites based on the Application's Phase (i.e. Login, Bug-Reporting). Currently defined suite information is located in the `protractor.base.conf.js` file as well.
#### Running E2E Tests

> E2E Tests can be executed by using `npm run e2e` which conducts testing  using the default `potractor.conf.js` in the Chrome Browser.   

Reviewer: ```suggestion
E2E Tests can be executed by using `npm run e2e` which conducts testing  using the default `protractor.conf.js` in the Chrome Browser.   
```

The following additional parameters would allow for further customisation,

| Additional Parameter | Description | Full Command Example |
| :---: | :-----: | :-------: |

> | `--potractor-config=e2e/potractor.*.conf.js` | Allows to substitute the default configuration file | `npm run e2e -- --potractor-config=e2e/potractor.firefox.conf.js` |

Reviewer: ```suggestion
| `--protractor-config=e2e/protractor.*.conf.js` | Allows to substitute the default configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` |
```

Author: Sure 👍

We loosely follow the [Jasmine Style Guide](https://github.com/CareMessagePlatform/jasmine-styleguide) when writing tests. 
One main guideline is that a `describe` block should be created for each method / scenario under test, and an `it` block should be created for each property being verified.

> ### E2E Testing

Reviewer 2: @ptvrajsk do move this PR's content into a new top-level section, thanks!

> ### E2E Testing

Author: Oh yes, missed this. 👍

> #### CATcher Test Prep

Reviewer 2: ```suggestion
#### How the E2E tests work
```

Let's change this heading to `How the E2E tests work` - it's easier to understand for new developers?

Author: Yes thats better 👍

E2E Tests are currently run using [Protractor](http://www.protractortest.org/#/) testing framework with the following stages. 

> 1. Build CATcher with using `test` architecture.

Reviewer 3: Grammatical mistake here!

```suggestion
1. Build CATcher using `test` architecture.
```

| Additional Parameter | Description | Full Command Example |
| :---: | :-----: | :-------: |

> | `--protractor-config=e2e/protractor.*.conf.js` | Allows substitution of the default configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` |

Reviewer 2: ```suggestion
| `--protractor-config=e2e/protractor.*.conf.js` | Allows substitution of the default configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` (runs the tests using firefox) |
```

Let's briefly explain what the example does?

Author: ![image](https://user-images.githubusercontent.com/39243082/110196039-c9b41980-7e7c-11eb-939d-aa265f6651b4.png)
Like this?

Reviewer 2: Yes

> | `--suite=*` | Runs E2E Tests for specific suites | `npm run e2e -- --suite=login,bugReporting`

Author: ![image](https://user-images.githubusercontent.com/39243082/110195951-1e0ac980-7e7c-11eb-8a5a-1ed1684229cb.png)
Is this better?

> | `--suite=*` | Runs E2E Tests for specific suites | `npm run e2e -- --suite=login,bugReporting`

Reviewer 2: Yea, there's a typo in `potractor.conf.js` though

## Troubleshooting conflicts between the versions of the browser and browser driver

If tests fail on your machine due to mismatches between the versions of the browser and the browser driver, you can use the [`webdriver-manager`](https://github.com/angular/webdriver-manager#readme) tool to install the right version of the driver.  By default, running `webdriver-manager update` updates all drivers to the latest version, but particular versions can be specified as options.
  
**TO NOTE:**

> - Relevant Browsers must be installed prior to running tests (i.e. Chrome, Firefox).

Reviewer 2: Let's move both these bullet points into the `Running E2E tests` section

>    - Data reflected in the environment file then assists the application replace the use of some existing services with those that bypass specific functions that are irrelevant to E2E Testing. This includes Authentication Bypassing, Backend API Simulation (so that tests can be carried out in isolation) and others.

Reviewer 3: ```suggestion
   - Data reflected in the environment file then assists the application in replacing the use of some existing services with those that bypass specific functions that are irrelevant to E2E Testing. This includes Authentication Bypassing, Backend API Simulation (so that tests can be carried out in isolation) and others.
```

>    - These Service Injections are carried in the respective `*-module.ts` files with the help of Factories (located in `/src/app/core/services/factories`) that check the current build environment and make the Service Replacements accordingly.

Reviewer 3: ```suggestion
   - These Service Injections are carried out in the respective `*-module.ts` files with the help of Factories (located in `/src/app/core/services/factories`) that check the current build environment and make the Service Replacements accordingly.
```

- CATcher can be launched in the Test Build by using the command `npm run ng:serve:test` to further develop mock services and debug E2E Tests.

## Protractor Configuration

- Protractor primarily requires the `*.conf.js` files to define E2E Testing Environments (this includes Browser Details, Base URL, etc...)
- The base configuration data is stored in `protractor.base.conf.js` which is then extended by separate configuration files for individual browsers as well as the CI/CD pipeline.
- E2E Tests are typically split into `Page-Objects Files` and `Test Files` in accordance with the [Protractor Style Guide](http://www.protractortest.org/#/style-guide) (more information regarding the interaction between the aforementioned filetypes can be found there).
- E2E Tests are also grouped into suites based on the Application's Phase (i.e. Login, Bug-Reporting). Currently defined suite information is located in the `protractor.base.conf.js` file as well.

## How the E2E tests work

> E2E Tests are currently run using [Protractor](http://www.protractortest.org/#/) testing framework with the following stages. 

Reviewer 2: ```suggestion
E2E Tests are run with the following stages:
```
Let's move the sentence `E2E Tests are currently run using [Protractor](http://www.protractortest.org/#/) testing framework` to the top of the `Protractor Configuration` section

> #### Running E2E Tests

Author: Sure I'll re-order them 👍

| Additional Parameter | Description | Full Command Example |
| :---: | :-----: | :-------: |

> | `--protractor-config=e2e/protractor.*.conf.js` | Allows to substitute the default configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` |

Reviewer 3: ```suggestion
| `--protractor-config=e2e/protractor.*.conf.js` | Allows substitution of the default configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` |
```

| `--suite=*` | Runs E2E Tests for specific suites | `npm run e2e -- --suite=login,bugReporting`

> #### webdriver-manager

Reviewer 2: ```suggestion
#### Troubleshooting conflicts between the versions of the browser and browser driver
```

Let's use a more descriptive heading (just `webdriver-manager` doesn't explain the purpose of the section)

Author: Ah yes sure 👍

> If tests are not correctly carried out in your local machine due to outdated Browser Drivers (e.g. ChromeDriver, GeckoDriver) a possible solution is to run `webdriver-manager update` which will attempt to update all local drivers to the latest version and should help mitigate any incompatibility issues. If you are still unable to run the E2E tests, please check that the Browser itself is up-to-date and re-run the webdriver command post-browser update.

Reviewer 2: ```suggestion
If tests fail on your machine due to mismatches between the versions of the browser and the browser driver, you can use the [`webdriver-manager`](https://github.com/angular/webdriver-manager#readme) tool to install the right version of the driver.  By default, running `webdriver-manager update` updates all drivers to the latest version, but particular versions can be specified as options.
```

Made these changes:
- Explicitly mention the use case for webdriver-manager
- Mentioned that webdriver-manager can install specific driver versions too. (on some machines, there could be a need to install an older browser driver, instead of *latest* version)
- Added link to the webdriver-manager's repo

  
**TO NOTE:**
- Relevant Browsers must be installed prior to running tests (i.e. Chrome, Firefox).

> - CATcher can be launched in the Test Build by using the command `npm run ng:serve:test` to furhter develop mock services and debug E2E Tests.

Reviewer 3: ```suggestion
- CATcher can be launched in the Test Build by using the command `npm run ng:serve:test` to further develop mock services and debug E2E Tests.
```

We loosely follow the [Jasmine Style Guide](https://github.com/CareMessagePlatform/jasmine-styleguide) when writing tests. 
One main guideline is that a `describe` block should be created for each method / scenario under test, and an `it` block should be created for each property being verified.

# E2E Testing

## Running E2E Tests

E2E Tests can be executed using `npm run e2e`. You should see CATcher launch in an instance of Google Chrome, with some automated actions occurring on it. Note: Google Chrome needs to be installed on the machine.

Unlike the production version of CATcher, we do not use the actual GitHub API in the E2E tests. Mock data is used to simulate the GitHub API. You can run `npm run ng:serve:test` to run CATcher in this "offline" mode (to further develop or debug the E2E tests).
The following additional parameters would allow for further customisation,

| Additional Parameter | Default | Description | Full Command Example | Command Explanation
| :---: | :---: | :-----: | :-------: | :------: |

> | `--protractor-config=e2e/protractor.*.conf.js` | `potractor.conf.js` | Allows selection of the Protractor configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` | Runs E2E Tests on the Firefox Browser |

Reviewer 2: ```suggestion
| `--protractor-config=e2e/protractor.*.conf.js` | `protractor.conf.js` | Allows selection of the Protractor configuration file | `npm run e2e -- --protractor-config=e2e/protractor.firefox.conf.js` | Runs E2E Tests on the Firefox Browser |
```

| `--suite=*` | All Suites | Runs E2E Tests for specific suites | `npm run e2e -- --suite=login,bugReporting` | Run E2E Tests from Login and BugReporting Suites only |

**TO NOTE:**
- Relevant Browsers must be installed prior to running tests (i.e. Chrome, Firefox).

> - CATcher can be launched in the Test Build by using the command `npm run ng:serve:test` to further develop mock services and debug E2E Tests.

Reviewer 2: Let's remove this sentence as this information is already mentioned above.

Author: Ah yes sure 👍
